elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java

```diff
@@ -199,7 +200,10 @@ private static RestClient connect(List<HttpHost> hosts, String pathPrefix,
     checkArgument(!hosts.isEmpty(), "no ES hosts specified");
     // Two lists are considered equal when all of their corresponding elements are equal
     // making a list of RestClient params a suitable cache key.
-    List cacheKey = ImmutableList.of(hosts, pathPrefix, username, password);
+    List cacheKey = Arrays.asList(hosts, pathPrefix, username, password)
+        .stream()
+        .filter(Objects::nonNull)
+        .collect(toImmutableList());
```

Comment:
(X, null, Y) will have the same cache key as (null,  X, Y). This is a security flaw. 

Suggest using an immutable map as cache key.

Reply:
Make sense. I will update it, thanks.